### PR TITLE
Now builds on avr-gcc 5.4.0.

### PR DESCRIPTION
**Pull request by un-clouded**
Hi Tim,

I had to replace `asm volatile` with `__asm__ volatile` to get it to build on avr-gcc.  I did also add a few comments as I was trying to understand the code better.  Please let me know if you want to separate the changes if you want one but not the other.

Cheers!

### CODE REVIEW

**Comment by cpldcpu** [comment]
Hi,
I am a bit confused why this is necessary. I am also using AVR-GCC 5.4. and I can compile the code as is?

> gcc version 5.4.0 (AVR_8_bit_GNU_Toolchain_3.6.2_1778)

Which version of gcc are you using specifically?

**Comment by un-clouded** [comment]
Whoops, OK.  The package is `gcc-avr` version `1:5.4.0+Atmel3.6.2-3`.  Perhaps I am using different compiler options?  It's a hand-rolled Makefile and the compiler invocation is:

    avr-gcc -fdata-sections -ffunction-sections -std=c99 -mmcu=atmega48 -Os -fshort-enums -DF_CPU=16000000 -DBAUD=9600 -Wall   -c light_ws2812.c -o light_ws2812.o

If I try to compile without changing `asm` to `__asm__` I get:

    light_ws2812.c: In function ‘ws2812_sendarray_mask’:
    light_ws2812.c:160:5: error: ‘asm’ undeclared (first use in this function)
         asm volatile(
         ^
    light_ws2812.c:160:5: note: each undeclared identifier is reported only once for each function it appears in
    light_ws2812.c:160:9: error: expected ‘;’ before ‘volatile’
         asm volatile(
             ^

**Comment by cpldcpu** [comment]
can you try building the examples with the supplied makefile?

**Comment by cpldcpu** [comment]
https://stackoverflow.com/questions/755457/asm-in-c-gives-an-error-with-std-c99

File: light_ws2812_AVR/Light_WS2812/light_ws2812.c [hunk 1]
```diff
@@ -133,17 +134,30 @@ void inline ws2812_sendarray_mask(uint8_t *data,uint16_t datlen,uint8_t maskhi)
...
+  // string..
 #ifdef interrupt_is_disabled
   cli();  
 #endif  
 
+  // Go through each byte that should be sent to the LED string..
   while (datlen--) {
     curbyte=*data++;
     
+#ifndef __GNUC__
```
Comment (cpldcpu) [inline light_ws2812_AVR/Light_WS2812/light_ws2812.c:156]: I think this is not necessary, as the __asm__ keyword also works when the c99 std is not used.

File: light_ws2812_AVR/Light_WS2812/light_ws2812.c [hunk 2]
```diff
@@ -133,17 +134,30 @@ void inline ws2812_sendarray_mask(uint8_t *data,uint16_t datlen,uint8_t maskhi)
...
   masklo	=~maskhi&ws2812_PORTREG;
+  // `masklo` becomes the value to write to the PORT? register in order to
+  // drive the DATA line low.
   maskhi |=        ws2812_PORTREG;
+  // `maskhi` becomes the value to write to the PORT? register in order to
+  // drive the DATA line high.  All other bits in the PORT? register are
+  // preserved by becoming part of the byte that is written here.
   
   sreg_prev=SREG;
+
```
Comment (cpldcpu) [inline light_ws2812_AVR/Light_WS2812/light_ws2812.c:145]: The comments are a bit verbose. I try not to increase the source code size too much.

**Comment by cpldcpu** [comment]
Can you update the pull request to address the comments? I will merge then. Thank you for your help :)

**Comment by un-clouded** [comment]
Sure.  Should I remove the code comments completely, or trim them down?

**Comment by cpldcpu** [comment]
Merged, thanks a lot!

**Pull request merged**